### ISSUE TITLE
Increasing batch size makes network forward x times slower

### ISSUE THREAD

**Issue by grigornalbandyan** [pytorch/pytorch#106485]
### 🐛 Describe the bug

I have a two layer network. The input is a 2D array of token ids, first layer is an embedding layer that replaces each pixel the respective embedding, the second layer does a convolution. Find the network definition below.
```python
class Model(nn.Module):
    def __init__(self, voc_dim, emb_dim, out_channels):
        super().__init__()
        self.embedding = nn.Embedding(
                num_embeddings=voc_dim,
                embedding_dim=emb_dim,
                padding_idx=0,
                sparse=True,
            )
        self.conv = nn.Conv2d(
                in_channels=emb_dim,
                out_channels=out_channels,
                kernel_size=3,
                stride=1,
                dilation=1,
                padding=1,
                bias=False,
            )
    def forward(self, x):
        # x is (batch_size, 672, 512), embeddings (batch_size, 768, 672, 512)
        embeddings = self.embedding(x).permute(0, 3, 1, 2)
        out = self.conv(embeddings)
        return out

network = Model(voc_dim=100000, emb_dim=768, out_channels=64)
network.cuda()
network.train()
```
For batch sizes of 4 and 8, inference with mixed precision, forward pass runs smoothly. But for batch size 16, forward becomes ~1000 times slower.  Batch size 16 does not fit in the memory without mixed precision. Forward pass is timed with the following code.
```python
def forward_speed(batch_size, n=5):
    forward_pass_timed = []
    for i in range(n):
        start = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        inp = torch.rand(batch_size, 672, 512).long().cuda()
        start.record()
        with torch.autocast(device_type="cuda", dtype=torch.float16):
            x = network(inp)
        end.record()
        torch.cuda.synchronize()
        forward_pass_timed.append(start.elapsed_time(end))
    print(f'Batch size - {batch_size}, Mean of {n} runs in ms {round(np.mean(forward_pass_timed), 2)}, STD {round(np.std(forward_pass_timed), 2)}')
```
The output of the function for batch sizes 4,8,16.
<img width="604" alt="Screenshot 2023-08-02 at 23 42 12" src="https://github.com/pytorch/pytorch/assets/127740186/c894d48b-1cee-408a-b85b-40878e71fd57">

Interestingly, when I measure just one layer i.e only embedding or only convolution, they are not slow, this effect is seen only when they are called together in a forward pass. GPU is A100 40GB.

### Versions

PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux 8.4 (Ootpa) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-18)
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.28

Python version: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-305.19.1.el8_4.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.0.221
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-PCIE-40GB
Nvidia driver version: 495.29.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              128
On-line CPU(s) list: 0-127
Thread(s) per core:  2
Core(s) per socket:  32
Socket(s):           2
NUMA node(s):        2
Vendor ID:           AuthenticAMD
CPU family:          25
Model:               1
Model name:          AMD EPYC 7543 32-Core Processor
Stepping:            1
CPU MHz:             1638.617
CPU max MHz:         2800.0000
CPU min MHz:         1500.0000
BogoMIPS:            5589.65
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            32768K
NUMA node0 CPU(s):   0-31,64-95
NUMA node1 CPU(s):   32-63,96-127
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate sme ssbd mba sev ibrs ibpb stibp vmmcall sev_es fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.0
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

**Comment by mikaylagawarecki** [comment]
Does this reproduce on 2.0.1?

I ran your script locally and got 

```
Batch size - 4, Mean of 10 runs in ms 183.35, STD 180.52
Batch size - 8, Mean of 10 runs in ms 247.35, STD 3.36
Batch size - 16, Mean of 10 runs in ms 517.17, STD 5.47
```

**Comment by grigornalbandyan** [comment]
Can this be because of CUDA driver or GPU specific bug? https://github.com/pytorch/pytorch/issues/106469 also report slowdown on A100.

**Comment by grigornalbandyan** [comment]
Benchmark with torch 2.0.1
<img width="595" alt="Screenshot 2023-08-04 at 00 28 08" src="https://github.com/pytorch/pytorch/assets/127740186/d4a7c310-5dd4-413f-949c-5bd924272a38">

**Comment by grigornalbandyan** [comment]
Are there any news in this matter @mikaylagawarecki ?

**Comment by awgu** [comment]
I can reproduce the error, and it seems like an issue for cudnn conv2d.

```
>>> import torch
>>> torch.backends.cudnn.version()
8401
```

@grigornalbandyan A temporary workaround can be to run `torch.backends.cudnn.enabled = False`. With that, I get the following times:
```
Batch size - 4, Mean of 10 runs in ms 232.69, STD 288.63
Batch size - 8, Mean of 10 runs in ms 274.14, STD 2.45
Batch size - 16, Mean of 1 runs in ms 1419.0, STD 0.0
```
Though, you are better off enabling cudnn and using batch size 8 from a performance perspective.

Also, if I run more than 1 iteration for batch size 16, I hit OOM on a 40 GB A100.

**Comment by grigornalbandyan** [comment]
@mikaylagawarecki I also reproduced the results with torch.backends.cudnn.enabled = False
```
Batch size - 4, Mean of 10 runs in ms 141.28, STD 13.39
Batch size - 8, Mean of 10 runs in ms 453.99, STD 214.06
Batch size - 16, Mean of 1 runs in ms 636.89, STD 0.0
```
Is there anything I can do to mitigate the issue but still have cudnn.enabled = True?

**Comment by cristianregep** [comment]
I see a  similar problem going on with my setup. Here's a minimal reproducible example

```python
import torch.nn.functional as F
import torch.nn as nn
import torch
from torch import Tensor
import time

class SoftmaxPooling(nn.Module):
    def __init__(self, num_features: int, pool_size: int):
        super().__init__()
        self.linear = nn.Linear(num_features, num_features, bias=False)
        self.pool_size = pool_size

    def forward(self, x: Tensor) -> Tensor:
        shapes = x.shape
        features = shapes[-2]
        x = x.transpose(-1, -2)
        x = self.linear(x)
        x = x.transpose(-1, -2)
        x = x.view(shapes[0], features, -1, self.pool_size)
        y = F.softmax(x, dim=-1)
        x = x * y
        x = torch.sum(x, dim=-1)
        return x

encoder = (
    nn.Sequential(
        nn.Conv1d(4, 768, 1),
        nn.BatchNorm1d(768),
        nn.GELU(),
        nn.Conv1d(768, 768, 1),
        SoftmaxPooling(768, 2),
        nn.BatchNorm1d(768),
        nn.GELU(),
        nn.Conv1d(768, 768, 1),
        SoftmaxPooling(768, 2),
    )
    .bfloat16()
    .cuda()
)
input_size = 256_000
vocab_size = 4
batch = torch.randint(0, 1, (25, vocab_size, input_size)).bfloat16().cuda()
```

The output times for the example above are (+ some warnings from CUDA):

> [/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306](https://vscode-remote+kubeflow-002eprod-002dsnow-002erelationsummer-002eco-002euk.vscode-resource.vscode-cdn.net/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306): UserWarning: Applied workaround for CuDNN issue, install nvrtc.so (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:84.)
>   return F.conv1d(input, weight, bias, self.stride,
> [/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306](https://vscode-remote+kubeflow-002eprod-002dsnow-002erelationsummer-002eco-002euk.vscode-resource.vscode-cdn.net/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306): UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
>   return F.conv1d(input, weight, bias, self.stride,
> Time taken 0: 33.29s
> Time taken 1: 35.45s
> Time taken 2: 33.17s
> Time taken 3: 35.54s
> Time taken 4: 33.22s

If you use the trick above, switching off cuDNN with `torch.backends.cudnn.enabled = False`, you get: 

> Time taken 0: 1.50s
> Time taken 1: 3.72s
> Time taken 2: 1.40s
> Time taken 3: 3.71s
> Time taken 4: 1.40s

You may find this useful, with `torch.backends.cudnn.enabled = True` and without bfloat16 you get reasonable numbers:

> [/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306](https://vscode-remote+kubeflow-002eprod-002dsnow-002erelationsummer-002eco-002euk.vscode-resource.vscode-cdn.net/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306): UserWarning: Applied workaround for CuDNN issue, install nvrtc.so (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:84.)
>   return F.conv1d(input, weight, bias, self.stride,
> [/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306](https://vscode-remote+kubeflow-002eprod-002dsnow-002erelationsummer-002eco-002euk.vscode-resource.vscode-cdn.net/opt/pyenv/versions/3.10.12/lib/python3.10/site-packages/torch/nn/modules/conv.py:306): UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
>   return F.conv1d(input, weight, bias, self.stride,
> Time taken 0: 2.85s
> Time taken 1: 3.64s
> Time taken 2: 3.98s
> Time taken 3: 3.91s
> Time taken 4: 3.94s

Maybe important, maybe not, but the reason I'm using Softmax pooling is due to Max pooling failing because of this bug https://github.com/pytorch/pytorch/issues/117736

**Environment**
PyTorch version: 2.3.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.12 (main, May  9 2024, 09:31:51) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-135-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 470.199.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          256
On-line CPU(s) list:             0-255
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7742 64-Core Processor
CPU family:                      23
Model:                           49
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       2
Stepping:                        0
Frequency boost:                 enabled
CPU max MHz:                     2250.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        4491.56
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate sme ssbd mba sev ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip rdpid overflow_recov succor smca
Virtualization:                  AMD-V
L1d cache:                       4 MiB (128 instances)
L1i cache:                       4 MiB (128 instances)
L2 cache:                        64 MiB (128 instances)
L3 cache:                        512 MiB (32 instances)
NUMA node(s):                    8
NUMA node0 CPU(s):               0-15,128-143
NUMA node1 CPU(s):               16-31,144-159
NUMA node2 CPU(s):               32-47,160-175
NUMA node3 CPU(s):               48-63,176-191
NUMA node4 CPU(s):               64-79,192-207
NUMA node5 CPU(s):               80-95,208-223
NUMA node6 CPU(s):               96-111,224-239
NUMA node7 CPU(s):               112-127,240-255
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy==1.10.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] pytorch-lightning==2.2.4
[pip3] pytorch_optimizer==2.12.0
[pip3] torch==2.3.0+cu118
[pip3] torch_geometric==2.5.3
[pip3] torchmetrics==1.4.0
[pip3] torchvision==0.18.0+cu118
[pip3] triton==2.3.0
[conda] Could not collect
